catboost/private/libs/algo/greedy_tensor_search.cpp

```diff
@@ -1385,62 +1459,170 @@ static TNonSymmetricTreeStructure GreedyTensorSearchDepthwise(
+                                                                      (TBucketStats* stats, TBucketStats* pstats,
+                                                                       TBucketStats* sstats, TIndexType id,
+                                                                       double& gainLocal,
+                                                                       const TCandidateInfo** bestSplitCandidateLocal,
+                                                                       TSplit& bestSplitLocal){
+            double bestScoreLocal = MINIMAL_SCORE; double scoreBeforeSplitLocal = 0;
+            CalcBestScoreLeafwise(data, {id}, ctx->LearnProgress->Rand.GenRand(), scoreStDev, &candidatesContexts, fold, ctx, stats, pstats, sstats, maxBucketCount, maxSplitEnsamples);
+            SelectBestCandidate(*ctx, candidatesContexts, maxFeatureValueCount, *fold, &bestScoreLocal, bestSplitCandidateLocal);
+            scoreBeforeSplitLocal = CalcScoreWithoutSplit(id, *fold, *ctx);
+            if(bestSplitCandidateLocal != nullptr) {
+                bestSplitLocal = (*bestSplitCandidateLocal)->GetBestSplit(
+                                            data,
+                                            *fold,
+                                            ctx->Params.CatFeatureParams->OneHotMaxSize);
+            }
+            gainLocal = bestScoreLocal - scoreBeforeSplitLocal;
+        };
+
+        TSplit bestSplitNext;
+        const TCandidateInfo* bestSplitCandidateNext = nullptr;
+        double nextGain = 0;
+
+        bool skip_next = false;
+
+        bool zero_sibling_left = false;
+        for (unsigned long id = 0; id < curLevelLeafs.size(); ++id) {
+            const auto& leafBounds = ctx->SampledDocs.LeavesBounds[curLevelLeafs[id]];
+
             if (leafBounds.GetSize() < ctx->Params.ObliviousTreeOptions->MinDataInLeaf) {
+                if(!skip_next) {
+                    skip_next = true;
+                    zero_sibling_left = true;
+                } else {
+                    skip_next = false;
+                }
                 continue;
             }
-            CalcBestScoreLeafwise(data, {leafToSplit}, ctx->LearnProgress->Rand.GenRand(), scoreStDev, &candidatesContexts, fold, ctx);
 
-            double bestScore = MINIMAL_SCORE;
             const TCandidateInfo* bestSplitCandidate = nullptr;
-            SelectBestCandidate(*ctx, candidatesContexts, maxFeatureValueCount, *fold, &bestScore, &bestSplitCandidate);
+
+            double gain = 0;
+
+            TSplit bestSplit;
+
+            if(curDepth == 0) {
+                TVector<TBucketStats> stats(maxBucketCount*nFeatures*maxSplitEnsamples);
+                CalcBestScoreAndCandidate(stats.data(),nullptr,nullptr,curLevelLeafs[id], gain, &bestSplitCandidate, bestSplit);
+                if (bestSplitCandidate == nullptr) {
+                    continue;
+                }
+                if (gain < 1e-9) {
```

Comment:
this occurs many times -- let us use a helper function
```c++
static void PushGain(...* parentsQueue, ...* stats) {
    constexpr double SmallGain = 1e-9;
    ...
}
```

Reply:
seems `continue` statement won't work as it's expected if we move it to separate function.

Reply:
done :)